flow-model-generator/modules/flow-model-generator-core/src/main/java/io/ballerina/flowmodelgenerator/core/ModuleNodeAnalyzer.java

```diff
@@ -0,0 +1,130 @@
+/*
+ *  Copyright (c) 2025, WSO2 LLC. (http://www.wso2.com)
+ *
+ *  WSO2 LLC. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.wso2.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package io.ballerina.flowmodelgenerator.core;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import io.ballerina.compiler.syntax.tree.DefaultableParameterNode;
+import io.ballerina.compiler.syntax.tree.FunctionDefinitionNode;
+import io.ballerina.compiler.syntax.tree.ModuleMemberDeclarationNode;
+import io.ballerina.compiler.syntax.tree.ModulePartNode;
+import io.ballerina.compiler.syntax.tree.Node;
+import io.ballerina.compiler.syntax.tree.NodeVisitor;
+import io.ballerina.compiler.syntax.tree.ParameterNode;
+import io.ballerina.compiler.syntax.tree.RequiredParameterNode;
+import io.ballerina.compiler.syntax.tree.RestParameterNode;
+import io.ballerina.compiler.syntax.tree.SeparatedNodeList;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.Token;
+import io.ballerina.flowmodelgenerator.core.model.ModuleInfo;
+import io.ballerina.flowmodelgenerator.core.model.NodeBuilder;
+import io.ballerina.flowmodelgenerator.core.model.NodeKind;
+import io.ballerina.flowmodelgenerator.core.model.Property;
+import io.ballerina.flowmodelgenerator.core.model.node.FunctionDefinitionBuilder;
+import io.ballerina.tools.text.LineRange;
+
+import java.util.Optional;
+
+/**
+ * Analyzes the module level functions and generates the flow model.
+ *
+ * @since 2.0.0
+ */
+public class ModuleNodeAnalyzer extends NodeVisitor {
+
+    private final ModuleInfo moduleInfo;
+    private final Gson gson;
+    private JsonElement node;
+
+    public ModuleNodeAnalyzer(ModuleInfo moduleInfo) {
+        this.moduleInfo = moduleInfo;
+        this.gson = new Gson();
+    }
+
+    public Optional<JsonElement> findFunction(ModulePartNode rootNode, String functionName) {
+        for (ModuleMemberDeclarationNode member : rootNode.members()) {
+            if (member.kind() == SyntaxKind.FUNCTION_DEFINITION) {
+                FunctionDefinitionNode functionNode = (FunctionDefinitionNode) member;
+                if (functionNode.functionName().text().equals(functionName)) {
+                    functionNode.accept(this);
+                    return Optional.of(this.node);
+                }
+            }
+        }
+        return Optional.empty();
+    }
```

Comment:
Here should we only expose public functions or, is it expected to return private functions as well?

Reply:
And how about special function types (e.g. arrow functions, external functions)?

Reply:
Before calling this API, VS Code must obtain the project components through the components API. At this point, it has information about both the file and function name. This API then scans the syntax tree of the specified file and returns **any function type** that matches the given function name. Any restrictions on the types of functions that should be provided should be handled by the components API